docs/flip/1009-optimizer-api.md

```diff
@@ -0,0 +1,453 @@
+
+Our current API (referred to as [previous API] in this document) uses a pattern
+where an `Optimizer` dataclass is created from a pytree of `target` variables
+and from an `OptimizerDef` that defines how to update optimizer state,
+hyperparameters, and target variables. This pattern is relatively complex for
+implementing a simple optimizer, while being quite verbose in the typical Linen
+train step (especially when using mutable state collections).
+
+This package `flax.optim` contains some optimizers, but that list is far from
+exhaustive and ideally we would instead use JAX optimizers from a dedicated PyPi
+package.
+
+DeepMind already has a dedicated library — [Optax] — that implements a wide
+range of interesting optimizers and provides a framework to compose new
+optimizers from reusable gradient transformations.
+
+[Optax]: https://github.com/deepmind/optax
+
+# Using Optax
+[Using Optax]: #using-optax
+
+## Gradient Transformations
+[Gradient Transformations]: #gradient-transformations
+
+While [Optax] does provide predefined optimizers (like `optax.adam`, or
+`optax.sgd` with momentum), it is really a library of *gradient transformations*
+and the idiomatic way of instantiating an optimizer is by providing a
+combination of these gradient transformations. To emulate the momentum
+optimizer from the example when using the [previous API] we would write:
+
+```python
+import optax
+
+tx = optax.chain(
+    optax.trace(decay=0.9, nesterov=False),
+    optax.scale_by_schedule(lambda step: -get_learning_rate(step)),
+)
+```
+
+Remarks:
+
+- Above gradient transformation would be equivalent with the example under
+  [Optimizer and OptimizerDef] where we define a Momentum optimizer without
+  Nesterov momentum (note that the `beta` parameter corresponds to the `decay`
+  parameter of the `optax.trace()` transformation, and the learning rate is
+  applied in a second chained transformation).
+- Note that hyper parameters like `decay` or `nesterov` only exist in the inner
+  scope of the higher order functions returning the `GradientTransformation`.
+  Such a gradient transformation is currently defined as a `NamedTuple` of the
+  `init()` and the `update()` function. In principle this pattern could be
+  extended to also store parameters, maybe a point to discuss on the [Optax]
+  repo.
+- We can use a `get_learning_rate()` that returns the learning rate depending on
+  the step number when defining the Optax gradient update transformation. Above
+  code illustrates how this can be a drop-in replacement for a function we also
+  use in our [previous training step], where this update function already exists
+  (notice how we need to invert the sign because we add the gradient update to
+  the parameters). In addition, you can use
+  [`schedule_hyperparams()`](https://github.com/deepmind/optax/pull/48) to
+  schedule arbitrary hyper parameters with Optax.
+
+## Optax Training Step
+[Optax Training Step]: #optax-training-step
+
+```python
```

Comment:
Since we aren't using the `variables` in this example, I guess we can simplify it to this, but I guess you put the variables in here to demonstrate how it works in general?

```python
@functools.partial(jax.jit, static_argnums=(4, 5))
def train_step(opt_state, params, inputs, labels, apply_fn, tx_update_fn):

  def loss_fn(params):
    logits = apply_fn({'params': params}, inputs, mutable=['batch_stats'])
    loss = xent_loss(logits, labels)
    return loss

  loss, grads = jax.value_and_grad(loss_fn, has_aux=True)(params)
  updates = tx_update_fn(grads, opt_state, params)
  new_params = optax.apply_updates(params, updates)
  return new_opt_state, new_params, loss

opt_state = tx.init(params)
for batch in ds.as_numpy_iterator():
  opt_state, params, loss = train_step(
      opt_state, params, batch['image'], batch['label'], model.apply,
      tx.update)
  print(loss)
```